### PR TITLE
Infer destination service resource adapt public api

### PR DESCRIPTION
**Pull request by stuartnelson3**
- Add support for exit spans, auto-infer their `destination.service.resource`
- Remove `destination.service.{name,type}` from the public api and mark them as deprecated

closes https://github.com/elastic/apm-agent-go/issues/976

### CODE REVIEW

File: spancontext.go [hunk 1]
```diff
@@ -57,10 +57,6 @@ type DatabaseSpanContext struct {
 
 // DestinationServiceSpanContext holds destination service span span.
 type DestinationServiceSpanContext struct {
-	// Name holds a name for the destination service, which may be used
-	// for grouping and labeling in service maps.
-	Name string
```
Comment (axw) [inline spancontext.go:62]: We can't remove this, that would be a breaking API change. Just mark it as deprecated and don't require it to be set.

File: spancontext.go [hunk 2]
```diff
@@ -222,10 +217,9 @@ func (c *SpanContext) SetMessage(message MessageSpanContext) {
 // Both service.Name and service.Resource are required. If either is empty,
 // then SetDestinationService is a no-op.
 func (c *SpanContext) SetDestinationService(service DestinationServiceSpanContext) {
-	if service.Name == "" || service.Resource == "" {
+	if service.Resource == "" {
 		return
 	}
-	c.destinationService.Name = truncateString(service.Name)
```
Comment (axw) [inline spancontext.go:228]: We still need to send this field to APM Server, as older versions (<7.14) require it. Seeing as we don't actually make use of the value, I think if it's unset we could just set it to `service.Resource`. Alternatively, see my comment about omitempty on the model type

File: validation_test.go [hunk 3]
```diff
@@ -119,12 +119,11 @@ func TestValidateDatabaseSpanContext(t *testing.T) {
 
 func TestValidateDestinationSpanContext(t *testing.T) {
 	overlong := strings.Repeat("x", 1025)
-	for _, name := range []string{"", overlong} {
+	for range []string{"", overlong} {
```
Comment (axw) [inline validation_test.go:122]: I think we should revert the changes to this file, given that we'll need to keep the API stable

File: gocontext.go [hunk 4]
```diff
@@ -90,6 +90,17 @@ func StartSpan(ctx context.Context, name, spanType string) (*Span, context.Conte
 	return StartSpanOptions(ctx, name, spanType, SpanOptions{})
 }
 
+// StartExitSpan starts an exit span. If the provided context already contains
+// an exit span, this function returns a dropped span.
```
Comment (axw) [inline gocontext.go:94]: instead of creating a new method, how about we add a `ExitSpan bool` field to `apm.SpanOptions`?

File: model/model.go [hunk 5]
```diff
@@ -371,10 +371,10 @@ type DestinationSpanContext struct {
 // DestinationServiceSpanContext holds contextual information about a
 // destination service.
 type DestinationServiceSpanContext struct {
-	// Type holds the destination service type.
+	// Type holds the destination service type. Deprecated.
 	Type string `json:"type,omitempty"`
 
-	// Name holds the destination service name.
+	// Name holds the destination service name. Deprecated.
 	Name string `json:"name,omitempty"`
```
Comment (axw) [inline model/model.go:378]: I forget whether we treat "" the same as omitted in apm-server, but maybe we can just remove "omitempty" from these two fields so we always send them as empty strings for backwards compatibility with older server versions that require the fields

File: span.go [hunk 6]
```diff
@@ -64,7 +76,8 @@ func (tx *Transaction) StartSpan(name, spanType string, parent *Span) *Span {
 // span type, subtype, and action; a single dot separates span type and
 // subtype, and the action will not be set.
 func (tx *Transaction) StartSpanOptions(name, spanType string, opts SpanOptions) *Span {
-	if tx == nil {
+	if tx == nil || opts.parent.IsExitSpan() {
+		// TODO: Is this the equivalent of a noop span?
```
Comment (axw) [inline span.go:80]: Yes, "dropped spans" are no-op spans. They exist to remove the need to perform nil-pointer checks all over the place when creating spans, mostly because spans may be dropped due to the max span limit being reached

File: span.go [hunk 7]
```diff
@@ -384,6 +400,30 @@ func (s *Span) ended() bool {
...
+func (s *Span) IsExitSpan() bool {
+	if s == nil || s.SpanData == nil {
+		return false
+	}
+	ctx := s.Context
+	return s.exit ||
+		ctx.destination != model.DestinationSpanContext{} ||
+		ctx.database != model.DatabaseSpanContext{} ||
+		ctx.message != model.MessageSpanContext{} ||
+		ctx.http != model.HTTPSpanContext{}
```
Comment (axw) [inline span.go:424]: Is this in the spec? I'd prefer not to second guess the "exit" field. If we do need to do this, then we'll need to lock the span -- s.SpanData can be modified concurrently or released back into a memory pool at any time.

File: span.go [hunk 8]
```diff
@@ -392,6 +432,7 @@ type SpanData struct {
 	stackTraceLimit        int
 	timestamp              time.Time
 	childrenTimer          childrenTimer
+	exit                   bool
```
Comment (axw) [inline span.go:435]: SpanData is removed when the span is ended, so I think we'll need to move `exit` to the Span type as a sibling to the Span.TraceContext field. That way we can still tell if a span is/was an exit span even after it's ended.

**Review by axw — CHANGES_REQUESTED** [review]

File: span.go (hunk 7 above)
Comment (stuartnelson3) [reply span.go:424]: Yeah, spec is available here: https://github.com/elastic/apm/blob/master/specs/agents/tracing-spans.md#exit-spans. Will add the locking.

Comment (axw) [reply span.go:424]: Okay, thanks. That's kind of a pain, because it means that when you end a span it'll lose it's inferred exit span status. I think we'd better set `span.exit=true` when ending the span if any of those properties are set, so we can persist the fact after it's ended.

File: span.go [hunk 9]
```diff
@@ -93,15 +81,20 @@ func (tx *Transaction) StartSpanOptions(name, spanType string, opts SpanOptions)
...
-		defer opts.parent.mu.Unlock()
 	}
 
 	// Prevent tx from being ended while we're starting a span.
 	tx.mu.RLock()
 	defer tx.mu.RUnlock()
 	if tx.ended() {
+		if opts.parent != nil {
+			opts.parent.mu.Unlock()
+		}
```
Comment (axw) [inline span.go:92]: Moving the unlock away from the lock scares me a bit. TBH I'm really not a fan of this implicit exit span status, so I've raised the question with agent devs whether it is necessary to implement that part of the spec.

Comment (stuartnelson3) [reply span.go:92]: :+1: makes sense. I'm not a fan of this locking either, so let's see what happens.

File: span.go [hunk 10]
```diff
@@ -384,6 +398,25 @@ func (s *Span) ended() bool {
 	return s.SpanData == nil
 }
 
+func (s *Span) setExitSpan(name, spanType string) {
```
Comment (axw) [inline span.go:401]: Looking at the spec, I believe the resource is meant to be set to `subtype || type`.

Subtype and Type can be set after starting a span, so I think we need to do that when ending the span. What we can do is keep track of whether SpanContext.SetDestinationService has been called (even with empty strings), and if it is never called for an exit span, then do this implicit call when ending the exit span.

**Review by axw — CHANGES_REQUESTED** [review]
Thanks, this is looking much better now. Just one issue now about timing of the implicit SetDestinationService call, and the values used.

File: span.go [hunk 11]
```diff
@@ -384,6 +403,24 @@ func (s *Span) ended() bool {
 	return s.SpanData == nil
 }
 
+func (s *Span) setExitSpan() {
```
Comment (axw) [inline span.go:406]: `setExitSpanDestinationService` might be a bit clearer on the intent?

**Review by axw — APPROVED** [review]
thanks, looks great :)
